### PR TITLE
Fix assets error

### PR DESCRIPTION
**Pull request by vivekvpandya**
## Issue
[CU-2dg6hd2](https://app.clickup.com/t/2dg6hd2) 

## Checklist

~- [] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_~
~- [] I have updated the `book/` to reflect changes made by this PR _(if applicable)_~

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2dg6hd2 Update checks and error messages in pallet-assets, pallet-bonded-finance](https://app.clickup.com/t/2dg6hd2)

File: runtime/primitives/src/currency.rs [hunk 1]
```diff
@@ -99,11 +103,28 @@ impl CurrencyId {
...
+	fn valid_currency_id(currency_id: CurrencyId) -> bool {
+		match currency_id {
+			CurrencyId::PICA |
+			CurrencyId::LAYR |
+			CurrencyId::CROWD_LOAN |
+			CurrencyId::KSM |
+			CurrencyId::kUSD |
+			CurrencyId::USDT |
+			CurrencyId::USDC => true,
+			_ => false,
```
Comment (hussein-aitlahcen) [inline runtime/primitives/src/currency.rs:123]: So an LP would be invalid... ?

**Review by hussein-aitlahcen — CHANGES_REQUESTED** [review]

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -846,6 +870,9 @@ pub mod pallet {
 				if asset == T::NativeAssetId::get() {
 					return <<T as Config>::NativeCurrency>::hold(who, amount)
 				}
+				if !T::ValidCurrency::valid_currency_id(asset) {
```
Comment (dzmitry-lahoda) [inline frame/assets/src/lib.rs:873]: ensure_foo_bar!(asset); or method, do not copy paste please

File: frame/lending/src/currency.rs [hunk 3]
```diff
@@ -194,6 +194,14 @@ pub mod defs {
...
 
 pub use defs::*;
+use primitives::currency::ValidCurrency;
 use scale_info::TypeInfo;
 
 pub type CurrencyId = u128;
+pub struct AllValidCurrencyId;
+impl ValidCurrency<CurrencyId> for AllValidCurrencyId {
+	fn valid_currency_id(_currency_id: CurrencyId) -> bool {
+		// all other assets in mock are valid
```
Comment (dzmitry-lahoda) [inline frame/lending/src/currency.rs:204]: I would prefer mocks reuse some trait from helpers and also do not allow 0 for consistency of tests and runtime

File: integration-tests/local-integration-tests/src/statemine.rs [hunk 4]
```diff
@@ -77,6 +77,7 @@ fn this_chain_statemine_transfers_back_and_forth_work() {
 	// but due to current half fee, sender asset should at lease: FEE_WEIGHT + 2 * FEE_KUSAMA
 	let relay_native_asset_amount = 3 * FEE_WEIGHT + 3 * FEE_KUSAMA;
 	let remote_asset_id = 3451561; // magic number to avoid zero defaults and easy to find
+							   // let remote_asset_id = 131;
```
Comment (dzmitry-lahoda) [inline integration-tests/local-integration-tests/src/statemine.rs:80]: ahah, sorry for this. please remove:)

File: frame/assets/src/lib.rs [hunk 5]
```diff
@@ -360,6 +367,14 @@ pub mod pallet {
 		}
 	}
 
+	pub(crate) fn valid_asset<T: Config>(asset_id: T::AssetId) -> Result<(), DispatchError> {
+		if T::ValidCurrency::validate(asset_id).is_err() {
+			Err(Error::<T>::InvalidCurrency.into())
+		} else {
+			Ok(())
+		}
+	}
```
[the remark this answers is not in the document]
Comment (benluelo) [reply frame/assets/src/lib.rs:376]: Is there a reason to have it return an error as opposed to a `bool`? you seem to check against `valid_asset().is_err()` most of the time

File: frame/assets/src/lib.rs [hunk 6]
```diff
@@ -98,7 +99,9 @@ pub mod pallet {
 	};
 	use frame_system::{ensure_root, ensure_signed, pallet_prelude::OriginFor};
 	use orml_traits::GetByKey;
-	use sp_runtime::DispatchError;
+	use primitives::currency::ValidateCurrencyId;
+	// use primitives::currency::{ValidCurrency, ValidateCurrencyId};
```
Comment (benluelo) [inline frame/assets/src/lib.rs:103]:
```suggestion
```

File: frame/assets/src/lib.rs [hunk 7]
```diff
@@ -115,6 +118,9 @@ pub mod pallet {
 		type WeightInfo: WeightInfo;
 		/// origin of admin of this pallet
 		type AdminOrigin: EnsureOrigin<Self::Origin>;
+		type Convert: Convert<u64, Self::Balance>;
+		// type ValidCurrency: ValidCurrency<Self::AssetId>;
```
Comment (benluelo) [inline frame/assets/src/lib.rs:122]:
```suggestion
```

**Review by PoisonPhang — CHANGES_REQUESTED** [review]
This wasn't the intended usage of the `Validate` type.
Ideally, you supply the function with a `Validated<YourType, (ValidationOne, ValidationTwo)> and then use `.value()` to get the actual type from the validated result. 
See [Vault](https://github.com/ComposableFi/composable/blob/main/frame/vault/src/lib.rs#L480) for an example.

File: runtime/primitives/src/currency.rs [hunk 8]
```diff
@@ -99,11 +100,46 @@ impl CurrencyId {
...
+			Asset { id: CurrencyId::USDC.0 as u64, name: b"USDC".to_vec() },
 		]
 		.to_vec()
 	}
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, TypeInfo)]
+pub struct ValidateCurrencyId;
+
+impl Validate<CurrencyId, ValidateCurrencyId> for ValidateCurrencyId {
```
Comment (benluelo) [inline runtime/primitives/src/currency.rs:113]: Possibly generalize to `T: Zero`?

**Review by benluelo — CHANGES_REQUESTED** [review]

File: frame/assets/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (benluelo) [reply frame/assets/src/lib.rs:376]: as @PoisonPhang mentioned, there's probably a better way to do this

[the remark this answers is not in the document]
Comment (PoisonPhang) [reply frame/assets/src/lib.rs:376]: Yeah, utilizing `Validated` will reduce a lot of this boilerplate and *should* remove the need for the `valid_asset` function.

File: frame/assets/src/lib.rs [hunk 9]
```diff
@@ -747,14 +773,17 @@ pub mod pallet {
 				if asset == T::NativeAssetId::get() {
 					return <<T as Config>::NativeCurrency>::set_balance(who, amount)
 				}
+				ValidatedAsset::<T>::new(asset).map_err(|_| Error::<T>::InvalidCurrency)?;
```
Comment (vivekvpandya) [inline frame/assets/src/lib.rs:776]: @PoisonPhang @benluelo is this better?

File: frame/assets/src/lib.rs [hunk 10]
```diff
@@ -956,20 +1001,32 @@ pub mod pallet {
 				if asset == T::NativeAssetId::get() {
 					return <<T as Config>::NativeCurrency>::total_issuance()
 				}
+				// if valid_asset::<T>(asset).is_err() {
+				// 	return T::Convert::convert(0_u64)
+				// }
+				if ValidatedAsset::<T>::new(asset).is_err() {
```
Comment (vivekvpandya) [inline frame/assets/src/lib.rs:1007]: and here too @PoisonPhang @benluelo ?

File: frame/assets/src/lib.rs (hunk 9 above)
Comment (PoisonPhang) [reply frame/assets/src/lib.rs:776]: Error handling looks better, but this still isn't correctly using Validated.
Something more like this is how Validated is intended to be used
```rust
// snip...
			fn set_balance(
				asset: Validated<Self::AssetId, ValidatedAsset>,
				who: &T::AccountId,
				amount: Self::Balance,
			) -> DispatchResult {
                let asset = asset.value;
				if asset == T::NativeAssetId::get() {
					return <<T as Config>::NativeCurrency>::set_balance(who, amount)
				}
                <<T as Config>::MultiCurrency>::set_balance(asset, who, amount)
             }
/ snip...
```
Again, see [this example](https://github.com/ComposableFi/composable/blob/main/frame/vault/src/lib.rs#L476) for the `amount` parameter from Vault

Comment (PoisonPhang) [reply frame/assets/src/lib.rs:776]: You may find the docs in [validation.rs](https://github.com/ComposableFi/composable/blob/main/frame/composable-support/src/validation.rs) helpful

File: frame/assets/src/lib.rs [hunk 11]
```diff
@@ -747,14 +756,17 @@ pub mod pallet {
...
 			}
 
 			fn set_total_issuance(asset: Self::AssetId, amount: Self::Balance) {
 				if asset == T::NativeAssetId::get() {
 					return <<T as Config>::NativeCurrency>::set_total_issuance(amount)
 				}
-				<<T as Config>::MultiCurrency>::set_total_issuance(asset, amount)
+				if let Some(asset) = valid_asset_id::<T>(asset) {
+					<<T as Config>::MultiCurrency>::set_total_issuance(asset, amount)
+				}
```
Comment (benluelo) [inline frame/assets/src/lib.rs:769]: What if invalid?

File: frame/dutch-auction/src/mock/runtime.rs [hunk 12]
```diff
@@ -154,6 +148,7 @@ impl pallet_assets::Config for Runtime {
 	type WeightInfo = ();
 	type AdminOrigin = EnsureSignedBy<RootAccount, AccountId>;
 	type GovernanceRegistry = GovernanceRegistry;
+	type ValidCurrency = ValidateCurrencyId;
```
Comment (benluelo) [inline frame/dutch-auction/src/mock/runtime.rs:151]: Perhaps rename this to `CurrencyValidator` or some other verb, to emphasize it's an action

File: frame/assets/src/lib.rs (hunk 11 above)
Comment (vivekvpandya) [reply frame/assets/src/lib.rs:769]: function signature does not return anything so if invalid do nothing.

**Review by benluelo — APPROVED** [review]

**Review by PoisonPhang — APPROVED** [review]

**Review by hussein-aitlahcen — APPROVED** [review]

**Pull request merged**